### PR TITLE
[Relay, Torch] Fix stack op axis check, support torch::stack conversion for a static list 

### PR DESCRIPTION
**Pull request by masahi**
This is a fix to the issue reported in https://github.com/apache/incubator-tvm/issues/6432

First, Relay stack op accepts `axis` arg:
https://github.com/apache/incubator-tvm/blob/082f27ebf8b14f537f0d7686e8161db1684f3110/python/tvm/topi/transform.py#L353-L354

Since this `axis` value refers to the axis in the stacked array, the following check is not correct. The axis value can be bigger by one than the input dimension. 
https://github.com/apache/incubator-tvm/blob/eee413f9d9f1157b777737adf39060dda1991841/src/relay/op/tensor/transform.cc#L298-L299

I believe the correct check should be 
```
  CHECK(-(ndim + 1) <= axis && axis < ndim + 1)
```

Second, our PyTorch frontend  currently only supports conversion of `torch::stack` for the dynamic input case. I added this conversion to support the following use case in LSTM.

https://github.com/apache/incubator-tvm/blob/50adbfac6a533f231a930bf80617c8eb3d7097a7/tests/python/frontend/pytorch/test_lstm.py#L77-L81

But since `torch::stack` with a static tensor list like `torch.stack((x, x))` is common, I added support for that here.

Please review @siju-samuel @t-vi @junrushao1994 @Laurawly

### CODE REVIEW

**Comment by masahi** [comment]
Windows build is failing, I don't know what to do about it @tqchen

**Comment by kevinthesun** [comment]
Thanks for fixing this. I'm trying to do the similar fix.

File: python/tvm/relay/frontend/pytorch.py [hunk 1]
```diff
@@ -1757,6 +1759,20 @@ def _impl(inputs, input_types):
...
 
+def _stack(prelude):
+    def _impl(inputs, input_types):
+        if isinstance(inputs[0], list):
+            # a static python list of tensors
+            dim = inputs[1]
+            return _op.stack(inputs[0], dim)
+        else:
+            # List ADT case
+            # TODO: is there a better way to check if an input is a List ADT?
```
Comment (zhiics) [inline python/tvm/relay/frontend/pytorch.py:1770]: let's put an assignee after TODO

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py:1770]: It seems if I do `infer_type` here, I get 
```
List[Tensor[(8, 8), float32]]
```

Also, if I do `prelude.mod.get_global_type_var("List")`, I get
```
List
```

so I'm looking for a way to test a type equality, any idea? cc @kevinthesun

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py:1770]: it seems this works

```
ty = _infer_type_with_prelude(inputs[0], prelude)
list_ty = prelude.mod.get_global_type_var("List")
assert ty.func == list_ty
            
```

Comment (masahi) [reply python/tvm/relay/frontend/pytorch.py]: A proper assertion added and TODO removed

**Review by t-vi — APPROVED** [review]
Looks great! Thanks @masahi .

**Review by siju-samuel — APPROVED** [review]
LGTM.

**Review by kevinthesun — APPROVED** [review]
LGTM
